### ISSUE TITLE
Order of eq? arguments can make occurrence typing less precise

### ISSUE THREAD

**Issue by jsmaniac** [racket/typed-racket#164]
Occurrence typing fails to narrow `(U (Pairof Nothing Any) (Pairof 'a Number))` down to `(Pairof 'a Number)` inside the if-branch of `(if (eq? 'a (car x)) ... ...)`, although it works for `(if (eq? (car x) 'a) ... ...)`, where the arguments to `eq?` were swapped.

It would be even better if the occurrence typing could eliminate the `(Pairof Nothing Any)` case directly, since it can never happen.

Tested on 6.2.900.5 and 6.2.900.6 (snapshot 20150728-75e19c9).

```scheme
#lang typed/racket

(: x (U (Pairof (U) Any)
        (Pairof 'a Number)))
(define x (cons 'a 1))

(if (eq? (car x) 'a)
    (+ (cdr x) 1)
    0)

(if (eq? 'a (car x)) ;; Swap the arguments to eq?
    (+ (cdr x) 1) ;; Error here, because type inference says x is (U (Pairof Nothing Any) (Pairof 'a Number))
    0)
```

**Comment by takikawa** [comment]
I looked into this a bit and I found out what is going wrong. TR looks at both sides of the `eq?` and sees if either is a type for a "flat" value like a string or symbol and branches on whether the left or right side is flat. The problem is that both sides can be, in which case it's not obvious which side to extract the refinement information from.

(the relevant lines are here https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-eq.rkt#L56)

I have a fix, but I'm going to think this through a bit more because I think there are more cases to consider. For example, what if the equality check refines a variable on *both* sides? I don't think TR has a way to express that.

**Comment by samth** [comment]
I think @andmkent probably has some thoughts here.

**Comment by andmkent** [comment]
1) @takikawa Could you expand on "what if the equality check refines a variable on both sides? I don't think TR has a way to express that." -- I'm not sure I understand. I assume you mean cases like (eq? x y), but I don't understand why that's difficult and/or problematic.

2) I could be missing something here... but can't we just do this in tc-app-eq.rkt:

```racket
...
(match* ((single-value v1) (single-value v2))
    [((tc-result1: (Value: (? ok? val1)) _ o1)
      (tc-result1: (Value: (? ok? val2)) _ o2))
     (ret -Boolean (-FS (-and (-filter (-val val2) o1)
                              (-filter (-val val1) o2))
                        (-and (-not-filter (-val val2) o1)
                              (-not-filter (-val val1) o2))))]
...
```

3) This point is not directly related to the issue we're bearing down on but... I think structural types w/ ⊥ as an accessible type should simplify to ⊥ as well... unless there's some reason we're not doing this already I'm not aware of.

**Comment by takikawa** [comment]
1-2) I think you're right, I was just being dumb.

**Comment by takikawa** [comment]
Closed by eb93a2b
